File: app/models/game.rb
class Game
  Dir.glob('app/game_commands/*.rb').each do |command_file|
    require_dependency command_file.match(/\/([a-z_]+)\.rb/)[1]
  end

  include Mongoid::Document
  include Mongoid::Timestamps
  extend MongoidIdentityMap

  field :name, :type => String
  field :turns_played, :type => Integer, :default => 0
  field :rounds_played, :type => Integer, :default => 0
  field :players, :type => Array, :default => []
  field :units, :type => String
  field :bases, :type => Array
  field :terrain_modifiers, :type => Array
  field :player_credits, :type => Array, :default => []
  field :starting_player_count, :type => Integer
  field :status, :type => String, :default => 'started'
  field :defeated_players, :type => Array, :default => []
  field :winner, :type => String
  field :chat_log, :type => Array, :default => []
  field :command_history, type: Array, default: []
  field :player_subscriptions, :type => Array, :default => []
  field :player_peace_offers, :type => Array, :default => []
  field :player_skips, type: Array, default: []
  field :turn_started_at, :type => DateTime
  field :time_limit, :type => Integer, :default => 24.hours.to_i
  field :reminder_sent_at, :type => DateTime
  field :game_type, :type => String, :default => 'subscriber'
  field :new_player, :type => Boolean, :default => false
  field :private, type: Boolean, default: false
  field :unrated, type: Boolean, default: false
  field :allow_replays, type: Boolean, default: true

  referenced_in :map
  #references_many :users, :stored_as => :array, :inverse_of => :games

  scope :visible_to_public, lambda { where(private: false) }
  scope :player_current, lambda { |u| where(:players => u._id, :status => 'started').not_in(:defeated_players => [u._id]) }
  scope :player_lost, lambda { |u| { :where => { :defeated_players => u._id } } }
  scope :player_won,  lambda { |u| { :where => { :winner => u._id.to_s } } }
  scope :player_finished, lambda { |u| { where: { players: u._id, :status.ne => 'started' } } }
  scope :available_to_player, lambda { |u|
    where(
      'this.players.length < this.starting_player_count'
    ).not_in(
      :players => [u.try(:_id)]
    ).where(
      :status => 'started'
    )
  }

  validates_presence_of :name, :allow_nil => false, :allow_blank => false

  before_create :setup_state
  before_create :increment_map_play_count
  before_save :serialize_units
  before_save :serialize_bases
  before_save :serialize_terrain_modifiers
  before_save :serialize_command_history

  def users
    @users ||= self.players.map { |p| User.find(p) }
  end

  def defeated_users
    @defeated_users ||= self.defeated_players.map { |p| User.find(p) }
  end

  def winner_user
    @winner ||= (winner ? User.find(self.winner) : nil)
  end

  def units
    @deserialized_units ||= (self.attributes[:units] and Marshal.load(self.attributes[:units].to_s))
  end

  def units=(units_array)
    @deserialized_units = units_array
  end

  def bases
    @deserialized_bases ||= (self.attributes[:bases] and Marshal.load(self.attributes[:bases].to_s))
  end

  def bases=(bases_array)
    @deserialized_bases = bases_array
  end

  def terrain_modifiers
    @deserialized_terrain_modifiers ||= (self.attributes[:terrain_modifiers] and Marshal.load(self.attributes[:terrain_modifiers].to_s))
  end

  def terrain_modifiers=(tms_array)
    @deserialized_terrain_modifiers = tms_array
  end

  def command_history
    if self.attributes[:command_history]
      @deserialized_command_history ||= self.attributes[:command_history].map { |c| Marshal.load(c.to_s) }
    end
  end

  def command_history=(command_array)
    @deserialized_command_history = command_array
  end

  def push_command!(command)
    self.command_history << command
    Game.collection.master.collection.update(
      { '_id' => self._id },
      { '$push' => { 'command_history' => BSON::Binary.new(Marshal.dump(command)) } }
    )
  end
  
  def current_user
    return self.users[0] if self.starting_player_count == 1
    self.users[self.turns_played % (self.starting_player_count)]
  end
 
  def creator
    self.users.first
  end

  def player_subscribed?(user)
    if user.try(:email_game_updates) and u = self.users.find { |u| u._id == user._id } and u = self.users.index(u)
      self.player_subscriptions[u]
    else
      false
    end
  end

  def update_player_subscription(user, sub)
    if u = self.users.find { |u| u._id == user._id } and u = self.users.index(u)
      self.player_subscriptions[u] = sub
      self.modify('player_subscriptions', nil, self.player_subscriptions)
      self.save
    end
  end

  def player_offered_peace?(user)
    if u = self.users.find { |u| u._id == user._id } and u = self.users.index(u)
      self.player_peace_offers[u]
    end
  end

  def update_player_peace_offer(user, peace)
    if u = self.users.find { |u| u._id == user._id } and u = self.users.index(u)
      self.player_peace_offers[u] = peace
      self.modify('player_peace_offers', nil, self.player_peace_offers)
      self.save
    end
  end

  def player_skip_count(user)
    if u = self.users.find { |u| u._id == user._id } and u = self.users.index(u)
      self.player_skips[u]
    end
  end

  def increment_player_skip_count(user)
    if u = self.users.find { |u| u._id == user._id } and u = self.users.index(u)
      self.player_skips[u] += 1
      self.modify('player_skips', nil, self.player_skips)
    end
  end

  def reset_player_skip_count(user)
    if u = self.users.find { |u| u._id == user._id } and u = self.users.index(u)
      self.player_skips[u] = 0
      self.modify('player_skips', nil, self.player_skips)
    end
  end
  
  def add_player!(user)
    if can_add_player?(user)
      self.users << user
      self.players << user._id
      self.modify('players', nil, self.players)
      self.turn_started_at = Time.now if self.current_user == user

      idx = self.players.size

      self.units.each do |u|
        u.player_id = user._id.to_s if u.player == idx
      end

      self.bases.each do |b|
        b.player_id = user._id.to_s if b.player == idx
      end

      self.game_type = user.account_type if self.players.size == 1

      self.save
    else
      false
    end
  end

  def can_add_player?(user)
    self.status == 'started' and \
    self.players.size < self.starting_player_count and \
    !self.players.include?(user._id) and \
    (!self.creator or self.game_type == 'free' or user.account_type == 'subscriber')
  end

  def remove_player!(user)
    if can_remove_player?(user)
      self.players.delete(user._id)
      self.save
    else
      false
    end
  end

  def can_remove_player?(user)
    ['waiting_for_players', 'all_present'].include?(self.status)
  end

  def player_number_for_user(user)
    self.players.index(user._id) + 1
  end

  def user_for_player_number(num)
    self.users[num - 1]
  end

  def user_credits(user)
    self.player_credits[self.players.index(user._id)] rescue nil
  end

  def set_user_credits(user, credits)
    self.player_credits[self.players.index(user._id)] = credits
  rescue
    nil
  end

  def set_player_credits(num, credits)
    self.player_credits[num - 1] = credits
  end

  def unit_owner(u)
    return nil if u['player'] == 0
    self.users[u['player'] - 1]
  end

  def base_owner(b)
    return nil if b['player'] == 0
    self.users[b['player'] - 1]
  end

  def send_chat_message(type, message, u = nil)
    message = { :msg_class => type, :message => message, :user_id => u.try(:id) }
    message.delete(:user_id) if !u
    self.append_to_chat_log(message)
    Orbited.send_data("game_#{self.id}", message.to_json)
  end

  def send_chat_message!(type, message, u = nil)
    self.send_chat_message(type, message, u)
    self.save
  end

  def append_to_chat_log(message)
    cl = self.chat_log.dup
    cl << message
    self.modify('chat_log', nil, cl)
  end
  
  def append_to_chat_log!(message)
    self.append_to_chat_log(message)
    self.save
  end

  def can_send_reminder?
    self.can_skip? and \
    (self.reminder_sent_at.nil? or (Time.now - self.reminder_sent_at) >= self.time_limit)
  end

  def can_skip?
    self.status == 'started' and self.current_user and \
    (self.turn_started_at.nil? or (Time.now - self.turn_started_at) >= self.time_limit)
  end
  
  def unit_at(x, y, slot = nil)
    slot = nil if slot == 'null'

    self.units.each do |u|
      if u.x == x and u.y == y
        return (slot ? u.loaded_units[slot] : u)
      end
    end

    return nil
  end

  def base_at(x, y)
    self.bases.each do |b|
      return b if b.x == x and b.y == y
    end

    return nil
  end

  def capturing_at?(x, y)
    if b = base_at(x, y)
      return true if b.capture_phase
    end

    return false
  end

  def rival_units
    self.units.select do |u|
      u.player_id != self.current_user._id.to_s
    end
  end

  def terrain_at(x, y)
    t = self.terrain_modifiers.reverse.find do |tm|
      true if tm.x == x and tm.y == y
    end

    if t
      return t.terrain_name.to_sym
    else
      return self.unmodified_terrain_at(x, y).to_sym
    end
  end

  def terrain_modifier_at(x, y)
    self.terrain_modifiers.reverse.find do |tm|
      true if tm.x == x and tm.y == y
    end
  end

  def unmodified_terrain_at(x, y)
    TileTypes[self.map.tiles[y][x]]
  end

  def current_round_command_pages
    # All pages of commands starting *after* the last EndTurn command
    # which was executed by the current player
    idx = self.command_history.size - 1
    cmds = self.command_history.reverse
    ret = {}

    cmds.each do |c|
      if self.current_user and c.includes_command_class?(EndTurn) and c.user.id == self.current_user.id
        break
      else
        ret[self.command_page_at_idx(idx)] = true
      end
      idx -= 1
    end

    ret.keys.each do |k|
      ret[k] = self.command_page(k)
    end

    ret
  end

  def current_round_command_pages_json
    ret = self.current_round_command_pages
    Hash[ret.map { |k, v| [k, v.map(&:to_json_hash)] }]
  end

  def command_page(p)
    first = p * 20
    last = first + 20
    self.command_history[first...last] || []
  end

  def command_page_at_idx(i)
    (i.to_f / 20.to_f).floor
  end

  def to_json_hash
    ret = self.attributes.slice(
      :_id, :map_id, :name, :starting_player_count, :status,
      :turns_played, :rounds_played, :player_credits,
      :player_subscriptions, :player_peace_offers, :player_skips,
      :turn_started_at, :time_limit, :reminder_sent_at,
      :chat_log
    ).merge(
      :map_tiles => self.map_tiles,
      :users => self.users.map(&:to_json_hash),
      :defeated_users => self.defeated_users.map(&:to_json_hash),
      :winner_user => self.winner_user,
      :units => self.units.map(&:to_json_hash),
      :bases => self.bases.map(&:to_json_hash),
      :terrain_modifiers => self.terrain_modifiers.try(:map, &:to_json_hash) || [],
      :paged_command_history => self.current_round_command_pages_json,
      :last_command_page => self.command_page_at_idx(self.command_history.size - 1),
      :preview_img => self.map.img_medium,
      :type => self.creator ? self.game_type : 'subscriber',
      :new_player => self.new_player
    )

    ret[:turn_started_at] = ret[:turn_started_at].to_i
    ret[:reminder_sent_at] = ret[:reminder_sent_at].try(:to_i) || 0
    ret
  end

  def to_json(opts = nil)
    if opts.nil?
      to_json_hash.to_json
    else
      super
    end
  end
  
  

  protected

  def map_tiles
    self.map.tiles
  end

  def player_owner_index(p)
    self.users.to_a.index(p) + 1
  end

  def setup_state
    us = map.units.dup
    bs = map.bases.dup
    tms = map.terrain_modifiers.dup rescue []

    player_indexes = (us.map { |u| u['player'] } + bs.map { |b| b['player'] }).uniq.sort - [0]

    us = us.map do |u|
      u = Unit.new(
        u['unit_type'].to_sym,
        (u['player'] == 0 ? nil : player_indexes.index(u['player']) + 1),
        u['x'], u['y']
      )
      u.summoning_sickness = false
      u.moved = false
      u
    end

    bs = bs.map do |b|
      Base.new(
        b['base_type'].to_sym,
        (b['player'] == 0 ? 0 : player_indexes.index(b['player']) + 1),
        b['x'], b['y']
      )
    end

    tms = tms.map do |tm|
      TerrainModifier.new(tm['terrain_name'], tm['x'], tm['y'])
    end

    self.units = us
    self.bases = bs
    self.terrain_modifiers = tms
    self.starting_player_count = map.player_count
    self.command_history = []

    self.player_credits = [map.starting_credits] * map.player_count

    unless self.player_credits.empty?
      self.player_credits[0] += self.bases.select do |b|
        b.base_type == :Base and b.player == 1
      end.size * 200
    end

    self.player_subscriptions = [true] * map.player_count
    self.player_peace_offers = [false] * map.player_count
    self.player_skips = [0] * map.player_count

    self.turn_started_at = Time.now
  end
  
  def increment_map_play_count
    Map.collection.master.collection.update(
      { '_id' => self.map_id },
      { '$inc' => { play_count: 1 } }
    )
  end

  def serialize_units
    self.modify('units', nil, BSON::Binary.new(Marshal.dump(@deserialized_units))) unless @deserialized_units.nil?
  end

  def serialize_bases
    self.modify('bases', nil, BSON::Binary.new(Marshal.dump(@deserialized_bases))) unless @deserialized_bases.nil?
  end

  def serialize_terrain_modifiers
    self.modify('terrain_modifiers', nil, BSON::Binary.new(Marshal.dump(@deserialized_terrain_modifiers))) unless @deserialized_terrain_modifiers.nil?
  end

  def serialize_command_history
    unless @deserialized_command_history.nil?
      arr = @deserialized_command_history.map { |c| BSON::Binary.new(Marshal.dump(c)) }
      self.modify('command_history', nil, arr)
    end
  end
end

File: public/javascripts/ec/game/game.js
var Game = new JS.Class({
  initialize: function(map_id, game_data) {
    this.id = game_data._id;
    this.map_id = game_data.map_id;
    this.name = game_data.name;
    this.starting_player_count = game_data.starting_player_count;
    this.status = game_data.status;
    this.turns_played = game_data.turns_played;
    this.rounds_played = game_data.rounds_played;
    this.chat_log = game_data.chat_log;
    this.preview_img = game_data.preview_img;
    this.type = game_data.type;
    this.new_player = game_data.new_player;

    this.players = [];
    for (var i in game_data.users) {
      this.players.push(new User(game_data.users[i]));
    }

    this._player_credits = game_data.player_credits;
    this.player_peace_offers = game_data.player_peace_offers;
    this.player_skips = game_data.player_skips;

    this.defeated_players = [];
    for (var i in game_data.defeated_users) {
      this.defeated_players.push(new User(game_data.defeated_users[i]));
    }

    this.winner = game_data.winner_user ? new User(game_data.winner_user) : null;

    this.map = new Map(map_id, game_data.map_tiles);
    this.map.draw();

    this.bases = new JS.Set([]);
    for (var i in game_data.bases) {
      var b = new Base(game_data.bases[i]);
      this.bases.add(b);
      b.add_to_map(this.map, b.x, b.y);
    }

    this.units = new JS.Set([]);
    for (var i in game_data.units) {
      var u = new Unit(game_data.units[i]);
      this.units.add(u);
      u.add_to_map(this.map, u.x, u.y);
    }

    this.terrain_modifiers = [];
    for (var i in game_data.terrain_modifiers) {
      var t = new TerrainModifier(game_data.terrain_modifiers[i]);
      this.add_terrain_modifier(t);
    }

    this.paged_command_history = {};
    for (var i in game_data.paged_command_history) {
      this.paged_command_history[i] = [];
      for (var j in game_data.paged_command_history[i]) {
        this.paged_command_history[i].push(this.new_game_command_for_game_from_json(game_data.paged_command_history[i][j]));
      }
    }
    this.last_command_page = game_data.last_command_page;
    this.command_page = this.last_command_page;

    if (this.command_page == -1) {
      this.command_page = 0;
      this.last_command_page = 0;
      this.paged_command_history[0] = [];
    }

    this.command_idx = this.paged_command_history[this.command_page].length - 1;

    this.player_subscriptions = game_data.player_subscriptions;
    this.time_limit = game_data.time_limit;

    this.turn_started_at = new Date();
    this.turn_started_at.setTime(game_data.turn_started_at * 1000);

    this.reminder_sent_at = new Date();
    this.reminder_sent_at.setTime(game_data.reminder_sent_at * 1000);
  },

  reset: function(callback) {
    var me = this;
    var game_loaded = false;
    var commands_loaded = false;
    var attempt_callback = function() {
      if (game_loaded && commands_loaded && callback) {
        callback();
      }
    }

    $.get('/maps/' + this.map_id + '.json', function(initial_map) {
      me.status = 'started';
      me.turns_played = 0;
      me.rounds_played = 0;

      me.defeated_players = [];
      me.winner = null;

      me._player_credits = [];
      for (var i = 1; i <= me.starting_player_count; i++) {
        me._player_credits[i] = me.map.starting_player_credits;
      }

      me.player_peace_offers = [];
      for (var i = 1; i <= me.starting_player_count; i++) {
        me.player_peace_offers.push(false);
      }

      me.player_skips = [];
      for (var i = 1; i <= me.starting_player_count; i++) {
        me.player_skips.push(0);
      }

      me.bases.forEach(function(b) { b.remove(); });
      me.bases = new JS.Set([]);
      for (var i in initial_map.bases) {
        var raw = initial_map.bases[i];
        var p = me.players[raw.player - 1];
        if (p) raw.player_id = p.id;

        var b = new Base(raw);
        me.bases.add(b);
        b.add_to_map(me.map, b.x, b.y);

        if (b.player == 1) { me._player_credits[0] += 200; }
      }

      me.units.forEach(function(u) { u.remove(); });
      me.units = new JS.Set([]);
      for (var i in initial_map.units) {
        var raw = initial_map.units[i];
        var p = me.players[raw.player - 1];
        if (p) raw.player_id = p.id;

        var u = new Unit(raw);
        me.units.add(u);
        u.add_to_map(me.map, u.x, u.y);
      }

      for (var i in me.terrain_modifiers) { me.terrain_modifiers[i].remove(); };
      me.terrain_modifiers = [];
      for (var i in initial_map.terrain_modifiers) {
        var raw = initial_map.terrain_modifiers[i];
        var tm = new TerrainModifier(raw);
        me.add_terrain_modifier(tm);
      }

      game_loaded = true;
      attempt_callback();
    });

    this.load_command_history_page(0, function() {
      me.command_idx = -1;
      me.command_page = 0;
      commands_loaded = true;
      attempt_callback();
    });
  },

  load_command_history_page: function(page, callback) {
    var me = this;

    if (!this.paged_command_history[page]) {
      $.get('/games/' + this.id + '/command_history.json', { 'page': page }, function(cmds) {
        me.paged_command_history[page] = [];

        for (var i in cmds.commands) {
          me.paged_command_history[page].push(me.new_game_command_for_game_from_json(cmds.commands[i]));
        }

        if (callback) callback();
      });
    } else {
      if (callback) callback();
    }
  },

  last_executed_command: function() {
    if (this.last_command_page == -1) return null;

    if (this.command_idx > -1) {
      return this.paged_command_history[this.command_page][this.command_idx];
    } else {
      if (this.command_page > 0) {
        var page = this.paged_command_history[this.command_page - 1];
        return page[page.length - 1];
      } else {
        return null;
      }
    }
  },

  pop_last_command: function() {
    if (this.command_idx == 0) {
      var ret = this.paged_command_history[this.last_command_page][0];

      if (this.last_command_page > 0) {
        this.paged_command_history[this.last_command_page] = null;
        this.command_page -= 1;
        this.last_command_page -= 1;
        this.command_idx = 19;
      } else {
        this.command_page -= 1;
        this.last_command_page -= 1;
        this.command_idx = -1;
      }

      return ret;
    } else {
      this.command_idx -= 1;
      return this.paged_command_history[this.last_command_page].pop();
    }
  },

  push_command: function(cmd) {
    if (this.last_command_page == -1) {
      this.last_command_page = 0;
      this.command_page = 0;
      this.paged_command_history[0] = [];
    }

    if (this.paged_command_history[this.last_command_page].length < 20) {
      this.paged_command_history[this.last_command_page].push(cmd);
      this.command_idx += 1;
    } else {
      this.last_command_page += 1;
      this.paged_command_history[this.last_command_page] = [];
      this.paged_command_history[this.last_command_page].push(cmd);
      this.command_idx = 0;
      this.command_page += 1;
    }
  },

  executed_last_command: function() {
    return (this.command_page == this.last_command_page && this.command_idx == this.paged_command_history[this.command_page].length - 1);
  },

  rewind: function(callback) {
    this.paged_command_history[this.command_page][this.command_idx].unexecute();
    this.command_idx -= 1;
    var me = this;

    if (this.command_idx < 0 && this.command_page > 0) {
      this.load_command_history_page(this.command_page - 1, function() {
        me.command_page -= 1;
        me.command_idx = me.paged_command_history[me.command_page].length - 1;
        if (callback) callback();
      });
    } else {
      if (callback) callback();
    }
  },

  rewind_until: function(until, callback) {
    var me = this;

    this.rewind(function() {
      if (!until() && me.last_executed_command()) {
        me.rewind_until(until, callback);
      } else {
        if (callback) callback();
      }
    });
  },

  fast_forward: function(callback) {
    this.command_idx += 1;
    var me = this;

    if (this.command_idx > this.paged_command_history[this.command_page].length - 1 && this.command_page < this.last_command_page) {
      this.load_command_history_page(this.command_page + 1, function() {
        me.command_idx = 0;
        me.command_page += 1;
        me.paged_command_history[me.command_page][me.command_idx].replay();
        if (callback) callback();
      });
    } else {
      this.paged_command_history[this.command_page][this.command_idx].replay();
      if (callback) callback();
    }
  },

  rewind_to_beginning: function(callback) {
    this.reset(callback);
  },

  rewind_to_last_round: function(callback) {
    var this_user = this.current_user();
    var this_round = this.rounds_played;
    if (this.status != 'started') this_round -= 1;
    var me = this;

    this.rewind_until(function() {
      return me.rounds_played < this_round && this_user.equals(me.current_user());
    }, callback);
  },

  step_backward: function(callback) {
    this.rewind(callback);
    return (this.command_page > 0 || this.command_idx > -1);
  },

  step_forward: function(callback) {
    this.fast_forward(callback);
    return (this.command_page < this.last_command_page || this.command_idx < this.paged_command_history[this.command_page].length - 1);
  },

  new_game_command_for_game_from_json: function(command_json) {
    var user = null;
    for (var j in this.players) {
      if (this.players[j].id == command_json.user_id) {
        user = this.players[j];
        break;
      }
    }

    return GameCommand.new_from_json(command_json, this, user);
  },

  current_user: function() {
    return this.players[this.turns_played % this.starting_player_count];
  },

  current_user_idx: function() {
    return (this.turns_played % this.starting_player_count) + 1;
  },

  user_idx: function(user) {
    for (var i = 0; i < this.players.length; i++) {
      if (this.players[i].id == user.id) {
        return i + 1;
      }
    }
  },

  add_player: function(u) {
    this.units.forEach(function(unit) {
      if (unit.player == this.players.length + 1) {
        unit.player_id = u.id;
      }
    }, this);

    this.bases.forEach(function(base) {
      if (base.player == this.players.length + 1) {
        base.player_id = u.id;
      }
    }, this);

    this.players.push(u);
  },

  unit_at: function(x, y, slot) {
    var u = this.units.find(function(unit) {
      return unit.x == x && unit.y == y;
    });

    if (slot != null) {
      return u.loaded_units[slot];
    } else {
      return u;
    }
  },

  friendly_unit_at: function(player, x, y) {
    var u = this.unit_at(x, y);

    if (u && u.player_id == player.id) {
      return u;
    } else {
      return false;
    }
  },

  enemy_units: function(player) {
    return new JS.Set(
      this.units.select(function(u) {
        return u.player_id != player.id;
      })
    );
  },

  base_at: function(x, y) {
    return this.bases.find(function(base) {
      return base.x == x && base.y == y;
    });
  },

  friendly_base_at: function(player, x, y) {
    var u = this.base_at(x, y);

    if (u && u.player_id == player.id) {
      return u;
    } else {
      return false;
    }
  },

  terrain_at: function(x, y) {
    var t = null;
    for (var i = this.terrain_modifiers.length - 1; i >= 0; i--) {
      var it = this.terrain_modifiers[i];
      if (it.x == x && it.y == y) {
        t = it;
        break;
      }
    }

    if (t) {
      return t.terrain_name;
    } else {
      return GameConfig.tiles[this.map.tile_index(x, y)];
    }
  },

  unmodified_terrain_at: function(x, y) {
    return GameConfig.tiles[this.map.tile_index(x, y)];
  },

  terrain_modifier_at: function(x, y) {
    for (var i = this.terrain_modifiers.length - 1; i >= 0; i--) {
      var it = this.terrain_modifiers[i];
      if (it.x == x && it.y == y) {
        return it;
      }
    }
  },

  add_terrain_modifier: function(t) {
    this.terrain_modifiers.push(t);
    t.add_to_map(this.map);
    this.update_modifiers_at(t.x, t.y);
  },

  remove_terrain_modifier: function(t) {
    for (var i in this.terrain_modifiers) {
      var tm = this.terrain_modifiers[i];
      if (tm.x == t.x && tm.y == t.y && tm.terrain_name == t.terrain_name) {
        this.terrain_modifiers.splice(i, 1);
        break;
      }
    }
    t.remove();
    this.update_modifiers_at(t.x, t.y);
  },

  update_modifiers_at: function(x, y) {
    this.update_terrain_modifier_for_surrounding_terrain(x, y);

    var surrounding = PathFinder.surrounding_tiles(new HCoordinate(x, y));
    for (var i in surrounding) {
      this.update_terrain_modifier_for_surrounding_terrain(surrounding[i].x, surrounding[i].y);
    }
  },

  update_terrain_modifier_for_surrounding_terrain: function(x, y) {
    var tm = this.terrain_modifier_at(x, y);
    if (tm) {
      var terrains = [];

      var surrounding = PathFinder.surrounding_tiles(new HCoordinate(x, y));
      for (var i in surrounding) {
        terrains.push(this.terrain_at(surrounding[i].x, surrounding[i].y));
      }

      tm.update_for_surrounding_terrain(terrains);
    }
  },

  can_build_bridge_at: function(x, y) {
    if ($.inArray(this.terrain_at(x, y), ['shallow_water', 'ford']) == -1) {
      return false;
    }

    var terrains = [];

    var surrounding = PathFinder.surrounding_tiles(new HCoordinate(x, y));
    for (var i in surrounding) {
      terrains.push(this.terrain_at(surrounding[i].x, surrounding[i].y));
    }

    return TmBridgeSprite.bridge_type_within(terrains) != null;
  },

  players_include: function(player) {
    for (var i in this.players) {
      if (this.players[i].equals(player)) {
        return true;
      }
    }

    return false;
  },

  defeated_players_include: function(player) {
    for (var i in this.defeated_players) {
      if (this.defeated_players[i].equals(player)) {
        return true;
      }
    }

    return false;
  },

  player_base_count: function(player) {
    return this.bases.select(function(base) {
      return base.player_id == player.id && base.base_type == 'Base';
    }, this).length;
  },

  player_credits: function(player) {
    var idx = 0;
    for (var i in this.players) {
      if (this.players[i].equals(player)) {
        idx = i; break;
      }
    }

    return this._player_credits[idx];
  },

  player_offered_peace: function(player) {
    var idx = 0;
    for (var i in this.players) {
      if (this.players[i].equals(player)) {
        idx = i; break;
      }
    }

    return this.player_peace_offers[idx];
  },

  player_skip_count: function(player) {
    var idx = 0;
    for (var i in this.players) {
      if (this.players[i].equals(player)) {
        idx = i; break;
      }
    }

    return this.player_skips[idx];
  },

  increment_player_skip_count: function(player) {
    var idx = 0;
    for (var i in this.players) {
      if (this.players[i].equals(player)) {
        idx = i; break;
      }
    }

    this.player_skips[idx] += 1;
  },

  reset_player_skip_count: function(player) {
    var idx = 0;
    for (var i in this.players) {
      if (this.players[i].equals(player)) {
        idx = i; break;
      }
    }

    this.player_skips[idx] = 0;
  },

  modify_player_credits: function(player, amt) {
    var idx = 0;
    for (var i in this.players) {
      if (this.players[i].equals(player)) {
        idx = i; break;
      }
    }

    this._player_credits[idx] += amt;
  },

  set_player_num_credits: function(player_num, amt) {
    this._player_credits[player_num - 1] = amt;
  },

  update_player_peace_offer: function(player, status) {
    var idx = 0;
    for (var i in this.players) {
      if (this.players[i].equals(player)) {
        idx = i; break;
      }
    }

    this.player_peace_offers[idx] = status;
  },

  modify_player_idx_credits: function(idx, amt) {
    this._player_credits[idx] += amt;
  },

  player_by_id: function(id) {
    for (var i in this.players) {
      if (this.players[i].id == id) return this.players[i];
    }
  },

  player_number: function(player) {
    for (var i in this.players) {
      if (this.players[i].equals(player)) {
        return parseInt(i) + 1;
      }
    }
  },

  player_by_number: function(player) {
    return this.players[player - 1];
  },

  can_undo: function() {
    if (this.last_command_page == -1) return false;

    if (this.paged_command_history[this.last_command_page].length > 0) {
      var last_command = this.last_executed_command();
      return last_command.can_undo();
    }

    return false;
  },

  over_time: function() {
    return this.can_skip();
  },

  can_skip: function() {
    var now = new Date().getTime();

    if (this.status == 'started' && this.current_user()) {
      if (this.turn_started_at == null || now - this.turn_started_at >= this.time_limit * 1000) {
        return true;
      }
    }

    return false;
  },

  can_send_reminder: function() {
    var now = new Date().getTime();

    if (this.can_skip()) {
      if (this.reminder_sent_at == null || now - this.reminder_sent_at >= this.time_limit * 1000) {
        return true;
      }
    }

    return false;
  }
});
